plugin/pkg/scheduler/algorithm/predicates/predicates.go

```diff
@@ -878,23 +878,26 @@ func haveSame(a1, a2 []string) bool {
 
 func GeneralPredicates(pod *v1.Pod, meta interface{}, nodeInfo *schedulercache.NodeInfo) (bool, []algorithm.PredicateFailureReason, error) {
 	var predicateFails []algorithm.PredicateFailureReason
-	fit, reasons, err := noncriticalPredicates(pod, meta, nodeInfo)
+	fit, reasons, err := EssentialPredicates(pod, meta, nodeInfo)
 	if err != nil {
 		return false, predicateFails, err
 	}
 	if !fit {
+		// No need check other predicates if the essential predicates failed
 		predicateFails = append(predicateFails, reasons...)
+		return len(predicateFails) == 0, predicateFails, nil
```

Comment:
Got it, thanks @wojtek-t , then we may need to check the failure reason twice to see if there are intentional reasons first, and do not fire InsufficientResourceError when there are intentional reasons. 

The following is an example of the PR, any comments for this approach?

```
--- a/pkg/controller/daemon/daemoncontroller.go
+++ b/pkg/controller/daemon/daemoncontroller.go
@@ -899,11 +899,35 @@ func (dsc *DaemonSetsController) nodeShouldRunDaemonPod(node *v1.Node, ds *exten
 		return false, false, false, err
 	}

+	// Do not fire InsufficientResourceError when there are intentional reasons
+	fireInsufficientResourceError := true
+	for _, r := range reasons {
+		switch reason := r.(type) {
+		case *predicates.PredicateFailureError:
+			// we try to partition predicates into two partitions here: intentional on the part of the operator and not.
+			switch reason {
+			// intentional
+			case
+				predicates.ErrNodeSelectorNotMatch,
+				predicates.ErrPodNotMatchHostName,
+				predicates.ErrNodeLabelPresenceViolated,
+				// this one is probably intentional since it's a workaround for not having
+				// pod hard anti affinity.
+				predicates.ErrPodNotFitsHostPorts,
+				// DaemonSet is expected to respect taints and tolerations
+				predicates.ErrTaintsTolerationsNotMatch:
+				fireInsufficientResourceError = false
+			}
+		}
+	}
+
 	for _, r := range reasons {
 		glog.V(4).Infof("DaemonSet Predicates failed on node %s for ds '%s/%s' for reason: %v", node.Name, ds.ObjectMeta.Namespace, ds.ObjectMeta.Name, r.GetReason())
 		switch reason := r.(type) {
 		case *predicates.InsufficientResourceError:
-			dsc.eventRecorder.Eventf(ds, v1.EventTypeWarning, FailedPlacementReason, "failed to place pod on %q: %s", node.ObjectMeta.Name, reason.Error())
+			if fireInsufficientResourceError {
+				dsc.eventRecorder.Eventf(ds, v1.EventTypeWarning, FailedPlacementReason, "failed to place pod on %q: %s", node.ObjectMeta.Name, reason.Error())
+			}
 			shouldSchedule = false
 		case *predicates.PredicateFailureError:
 			var emitEvent bool
```